neonvm/main.go

```diff
@@ -95,48 +106,58 @@ func main() {
 		// LeaderElectionReleaseOnCancel: true,
 	})
 	if err != nil {
-		setupLog.Error(err, "unable to start manager")
-		os.Exit(1)
+		logPanic(err, "unable to start manager")
 	}
 
 	if err = (&controllers.VirtualMachineReconciler{
 		Client:   mgr.GetClient(),
 		Scheme:   mgr.GetScheme(),
 		Recorder: mgr.GetEventRecorderFor("virtualmachine-controller"),
 	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "VirtualMachine")
-		os.Exit(1)
+		logPanic(err, "unable to create controller", "controller", "VirtualMachine")
 	}
 	if err = (&vmv1.VirtualMachine{}).SetupWebhookWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create webhook", "webhook", "VirtualMachine")
-		os.Exit(1)
+		logPanic(err, "unable to create webhook", "webhook", "VirtualMachine")
 	}
 	if err = (&controllers.VirtualMachineMigrationReconciler{
 		Client:   mgr.GetClient(),
 		Scheme:   mgr.GetScheme(),
 		Recorder: mgr.GetEventRecorderFor("virtualmachinemigration-controller"),
 	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "VirtualMachineMigration")
-		os.Exit(1)
+		logPanic(err, "unable to create controller", "controller", "VirtualMachineMigration")
 	}
 	if err = (&vmv1.VirtualMachineMigration{}).SetupWebhookWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create webhook", "webhook", "VirtualMachineMigration")
-		os.Exit(1)
+		logPanic(err, "unable to create webhook", "webhook", "VirtualMachineMigration")
 	}
 	//+kubebuilder:scaffold:builder
 
 	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up health check")
-		os.Exit(1)
+		logPanic(err, "unable to set up health check")
 	}
 	if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
-		setupLog.Error(err, "unable to set up ready check")
-		os.Exit(1)
+		logPanic(err, "unable to set up ready check")
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
+	defer cancel()
+	ctx = srv.SetShutdownSignal(ctx)
+	ctx = srv.SetBaseContext(ctx)
+	ctx = srv.WithOrchestrator(ctx)
+	orca := srv.GetOrchestrator(ctx)
+	defer func() {
+		if err := orca.Wait(); err != nil {
+			logPanic(err, "Failed to shut down orchestrator")
+		}
+
+		setupLog.Info("Main loop returned. Exiting.")
+	}()
+
+	if err := orca.Add(srv.HTTP("scheduler-pprof", time.Second, util.MakePPROF("0.0.0.0:7777"))); err != nil {
+		logPanic(err, "Failed to add pprof service")
 	}
 
 	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
-		os.Exit(1)
+	if err := mgr.Start(ctx); err != nil {
+		logPanic(err, "problem running manager")
```

Comment:
These lines will result in artificially increased severity for problems encountered at runtime.

For example, [errors from failed leadership election](https://neondb.slack.com/archives/C03F5SM1N02/p1701819152996709?thread_ts=1701817648.282449&cid=C03F5SM1N02) would result in panicking here, which - when someone reads the logs - are likely to be interpreted as a more significant problem than actually exists (especially because the stack trace of the panic does not add useful information about the origin of the error).

A more complete solution would probably involve something like how the scheduler sets up a "real" main that holds the deferred functions and is allowed return error:

https://github.com/neondatabase/autoscaling/blob/9b1e55c7cb55f503223d9817930a387e8622f297/cmd/autoscale-scheduler/main.go#L25-L38

Reply:
Resolved in c89494ff5ffc7f996e415aaa283d36350f245531. I decided to keep the general structure of main.go intact since it's generated by kubebuilder and has some markers for future code insertions. So, I didn't feel good changing it.